Assisted factory is generated for unsupported variants

## Description:
The following code results in the generation of `PhoneAssistedFactoryImpl` inside `TabletComponent`, with an `Optional` in the constructor, which then leads to a compilation error:

```
@AssistedFactory
interface PhoneAssistedFactory {
    fun create(i: Int): PhoneAssistedClass
}

@Conditional(onlyIn = [DeviceType.Phone::class])
class PhoneAssistedClass @AssistedInject constructor(
    phoneDependency: PhoneDependency,
    @Assisted arg: Int,
)

@Conditional(onlyIn = [DeviceType.Phone::class])
class PhoneDependency @Inject constructor()

class CommonClass @Inject constructor(
    phoneFactory: Optional<PhoneAssistedFactory>,
)


interface CommonComponent {
    fun getCommonClass(): CommonClass
}

@Component(variant = [DeviceType.Phone::class])
interface PhoneComponent: CommonComponent

@Component(variant = [DeviceType.Tablet::class])
interface TabletComponent: CommonComponent
```

```
/Users/newuserkk/projects/newuserkk/yatagan-assisted-factory-reproducer/app/build/generated/source/kapt/debug/com/newuserkk/yatagan/Yatagan$TabletComponent.java:32: error: incompatible types: no instance(s) of type variable(s) T exist so that Optional<T> conforms to PhoneDependency
      return new PhoneAssistedClass(Optional.empty(), i);
             ^
  where T is a type-variable:
    T extends Object declared in method <T>empty()
```

## Expected result
An assisted factory should probably not be generated for an unsupported variant component at all

## Environment
Yatagan 1.6.1 with kapt

## Minimal reproducer
https://github.com/NewUserKK/yatagan-assisted-factory-reproducer

--- a/.gitignore
+++ b/.gitignore
@@ -1,41 +1,8 @@
-# Covers JetBrains IDEs: IntelliJ, RubyMine, PhpStorm, AppCode, PyCharm, CLion, Android Studio, WebStorm and Rider
-# Reference: https://intellij-support.jetbrains.com/hc/en-us/articles/206544839
-# From: https://github.com/github/gitignore/blob/master/Global/JetBrains.gitignore
-
 # User-specific stuff
-.idea/**/workspace.xml
-.idea/**/tasks.xml
-.idea/**/usage.statistics.xml
-.idea/**/dictionaries
-.idea/**/shelf
-
-# Generated files
-.idea/**/contentModel.xml
-
-# Sensitive or high-churn files
-.idea/**/dataSources/
-.idea/**/dataSources.ids
-.idea/**/dataSources.local.xml
-.idea/**/sqlDataSources.xml
-.idea/**/dynamic.xml
-.idea/**/uiDesigner.xml
-.idea/**/dbnavigator.xml
-
-# Gradle
-.idea/**/gradle.xml
-.idea/**/libraries
-
-# Gradle and Maven with auto-import
-# When using Gradle or Maven with auto-import, you should exclude module files,
-# since they will be recreated, and may cause churn.
-.idea/artifacts
-.idea/compiler.xml
-.idea/jarRepositories.xml
-.idea/modules.xml
-.idea/*.iml
-.idea/modules
-*.iml
-*.ipr
+.idea/
+!.idea/codeStyles
+!.idea/inspectionProfiles
+!.idea/vcs.xml
 
 # Build dirs
 /build/

--- a/.idea/kotlinc.xml
+++ b/.idea/kotlinc.xml
@@ -1,6 +1,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<project version="4">
-  <component name="KotlinJpsPluginSettings">
-    <option name="version" value="1.9.24" />
-  </component>
-</project>

--- a/.idea/misc.xml
+++ b/.idea/misc.xml
@@ -1,7 +1,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<project version="4">
-  <component name="ExternalStorageConfigurationManager" enabled="true" />
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" project-jdk-name="11" project-jdk-type="JavaSDK">
-    <output url="file://$PROJECT_DIR$/out" />
-  </component>
-</project>

--- a/codegen/impl/src/main/kotlin/AssistedInjectFactoryGenerator.kt
+++ b/codegen/impl/src/main/kotlin/AssistedInjectFactoryGenerator.kt
@@ -24,6 +24,7 @@ import com.yandex.yatagan.codegen.poetry.buildExpression
 import com.yandex.yatagan.core.graph.BindingGraph
 import com.yandex.yatagan.core.graph.bindings.AssistedInjectFactoryBinding
 import com.yandex.yatagan.core.model.AssistedInjectFactoryModel
+import com.yandex.yatagan.core.model.ConditionScope
 import com.yandex.yatagan.core.model.component1
 import com.yandex.yatagan.core.model.component2
 import javax.inject.Inject
@@ -37,13 +38,15 @@ internal class AssistedInjectFactoryGenerator @Inject constructor(
 ) : ComponentGenerator.Contributor {
     private val modelToImpl: Map<AssistedInjectFactoryModel, ClassName> by lazy {
         val classNamespace = Namespace()
-        thisGraph.localAssistedInjectFactories.associateWith { model ->
-            componentImplName.nestedClass(classNamespace.name(
-                nameModel = model.name,
-                suffix = "Impl",
-                firstCapital = true,
-            ))
-        }
+        thisGraph.localAssistedInjectFactories
+            .filter { it.value != ConditionScope.Never }
+            .mapValues {
+                componentImplName.nestedClass(classNamespace.name(
+                    nameModel = it.key.name,
+                    suffix = "Impl",
+                    firstCapital = true,
+                ))
+            }
     }
 
     fun generateCreation(

--- a/core/graph/api/src/main/kotlin/BindingGraph.kt
+++ b/core/graph/api/src/main/kotlin/BindingGraph.kt
@@ -65,7 +65,7 @@ public interface BindingGraph : MayBeInvalid, Extensible, WithParents<BindingGra
     /**
      * [AssistedInjectFactoryModel]s that are hosted in this graph.
      */
-    public val localAssistedInjectFactories: Collection<AssistedInjectFactoryModel>
+    public val localAssistedInjectFactories: Map<AssistedInjectFactoryModel, ConditionScope>
 
     /**
      * A collection of parent (not necessarily direct) [BindingGraph]s, from which bindings and/or conditions are used

--- a/core/graph/impl/src/main/kotlin/BindingGraphImpl.kt
+++ b/core/graph/impl/src/main/kotlin/BindingGraphImpl.kt
@@ -121,7 +121,7 @@ internal class BindingGraphImpl(
     internal val localNodes = mutableSetOf<NodeModel>()
     override val localBindings = mutableMapOf<Binding, BindingUsageImpl>()
     override val localConditionLiterals = mutableMapOf<ConditionModel, LiteralUsage>()
-    override val localAssistedInjectFactories = mutableSetOf<AssistedInjectFactoryModel>()
+    override val localAssistedInjectFactories = mutableMapOf<AssistedInjectFactoryModel, ConditionScope>()
     override val usedParents = mutableSetOf<BindingGraph>()
     override val children: Collection<BindingGraphImpl>
 
@@ -208,7 +208,7 @@ internal class BindingGraphImpl(
             binding.accept(object : BindingVisitorAdapter<Unit>() {
                 override fun visitDefault() = Unit
                 override fun visitAssistedInjectFactory(binding: AssistedInjectFactoryBinding) {
-                    localAssistedInjectFactories += binding.model
+                    localAssistedInjectFactories[binding.model] = binding.conditionScope
                 }
             })
         }
@@ -256,7 +256,7 @@ internal class BindingGraphImpl(
         for (child in childrenSequence(includeThis = false)) {
             child as BindingGraphImpl
             val usesConditions = child.localConditionLiterals.keys.removeAll(localConditionLiterals.keys)
-            val usesAssistedInjectFactories = child.localAssistedInjectFactories.removeAll(localAssistedInjectFactories)
+            val usesAssistedInjectFactories = child.localAssistedInjectFactories.keys.removeAll(localAssistedInjectFactories.keys)
             if (usesConditions || usesAssistedInjectFactories) {
                 // This will never be seen by materialization and that's okay, because no bindings are required here.
                 child.usedParents += this

--- a/testing/tests/src/test/kotlin/ConditionsTest.kt
+++ b/testing/tests/src/test/kotlin/ConditionsTest.kt
@@ -1219,4 +1219,52 @@ class ConditionsTest(
 
         compileRunAndValidate()
     }
+
+    @Test
+    fun `issue #209 - assisted factory generated for unsupported variants`() {
+        includeFromSourceSet(flavors)
+
+        givenKotlinSource("test.TestCase", """
+            import com.yandex.yatagan.*
+            import javax.inject.*
+
+            @AssistedFactory
+            interface PhoneAssistedFactory {
+                fun create(i: Int): PhoneAssistedClass
+            }
+
+            @Conditional(onlyIn = [DeviceType.Phone::class])
+            class PhoneAssistedClass @AssistedInject constructor(
+                phoneDependency: PhoneDependency,
+                @Assisted arg: Int,
+            )
+
+            @Conditional(onlyIn = [DeviceType.Phone::class])
+            class PhoneDependency @Inject constructor()
+
+            class CommonClass @Inject constructor(
+                phoneFactory: Optional<PhoneAssistedFactory>,
+            )
+
+            interface CommonComponent {
+                fun getCommonClass(): CommonClass
+            }
+
+            @Component(variant = [DeviceType.Phone::class])
+            interface PhoneComponent: CommonComponent
+
+            @Component(variant = [DeviceType.Tablet::class])
+            interface TabletComponent: CommonComponent
+
+            fun test() {
+                val phoneComponent = Yatagan.create(PhoneComponent::class.java)
+                val tabletComponent = Yatagan.create(TabletComponent::class.java)
+
+                val phoneCommon = phoneComponent.getCommonClass()
+                val tabletCommon = tabletComponent.getCommonClass()
+            }
+        """.trimIndent())
+
+        compileRunAndValidate()
+    }
 }

--- a/testing/tests/src/test/resources/golden/ConditionsTest/issue-209-assisted-factory-generated-for-unsupported-variants.golden-code.txt
+++ b/testing/tests/src/test/resources/golden/ConditionsTest/issue-209-assisted-factory-generated-for-unsupported-variants.golden-code.txt
@@ -1,0 +1,102 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Name: test/Yatagan$TabletComponent.java
+package test;
+
+import com.yandex.yatagan.AutoBuilder;
+import com.yandex.yatagan.Optional;
+import com.yandex.yatagan.internal.Checks;
+import com.yandex.yatagan.internal.YataganGenerated;
+import java.lang.Class;
+import java.lang.Override;
+import java.lang.SuppressWarnings;
+import java.util.Collections;
+import javax.annotation.processing.Generated;
+
+@SuppressWarnings({"unchecked", "rawtypes", "NullableProblems"})
+@YataganGenerated
+@Generated("com.yandex.yatagan.codegen.impl.ComponentGenerator")
+public final class Yatagan$TabletComponent implements TabletComponent {
+  private Yatagan$TabletComponent() {
+  }
+
+  @Override
+  public CommonClass getCommonClass() {
+    return new CommonClass(Optional.empty());
+  }
+
+  public static AutoBuilder<Yatagan$TabletComponent> autoBuilder() {
+    return new AutoBuilderImpl();
+  }
+
+  private static final class AutoBuilderImpl implements AutoBuilder<Yatagan$TabletComponent> {
+    @Override
+    public final <I> AutoBuilder<Yatagan$TabletComponent> provideInput(I input,
+        Class<I> inputClass) {
+      Checks.reportUnexpectedAutoBuilderInput(input.getClass(), Collections.emptyList());
+      return this;
+    }
+
+    @Override
+    public final Yatagan$TabletComponent create() {
+      return new Yatagan$TabletComponent();
+    }
+  }
+}
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Name: test/Yatagan$PhoneComponent.java
+package test;
+
+import com.yandex.yatagan.AutoBuilder;
+import com.yandex.yatagan.Optional;
+import com.yandex.yatagan.internal.Checks;
+import com.yandex.yatagan.internal.YataganGenerated;
+import java.lang.Class;
+import java.lang.Override;
+import java.lang.SuppressWarnings;
+import java.util.Collections;
+import javax.annotation.processing.Generated;
+
+@SuppressWarnings({"unchecked", "rawtypes", "NullableProblems"})
+@YataganGenerated
+@Generated("com.yandex.yatagan.codegen.impl.ComponentGenerator")
+public final class Yatagan$PhoneComponent implements PhoneComponent {
+  private Yatagan$PhoneComponent() {
+  }
+
+  @Override
+  public CommonClass getCommonClass() {
+    return new CommonClass(this.optOfPhoneAssistedFactory());
+  }
+
+  Optional optOfPhoneAssistedFactory() {
+    return Optional.of(this.new PhoneAssistedFactoryImpl());
+  }
+
+  public static AutoBuilder<Yatagan$PhoneComponent> autoBuilder() {
+    return new AutoBuilderImpl();
+  }
+
+  private final class PhoneAssistedFactoryImpl implements PhoneAssistedFactory {
+    @Override
+    public PhoneAssistedClass create(int i) {
+      return new PhoneAssistedClass(new PhoneDependency(), i);
+    }
+  }
+
+  private static final class AutoBuilderImpl implements AutoBuilder<Yatagan$PhoneComponent> {
+    @Override
+    public final <I> AutoBuilder<Yatagan$PhoneComponent> provideInput(I input,
+        Class<I> inputClass) {
+      Checks.reportUnexpectedAutoBuilderInput(input.getClass(), Collections.emptyList());
+      return this;
+    }
+
+    @Override
+    public final Yatagan$PhoneComponent create() {
+      return new Yatagan$PhoneComponent();
+    }
+  }
+}
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

